File: January/Java/Pseudo-Palindromic Paths in a Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    static Map<Integer,Integer> map;
    static int count=0;
    public int pseudoPalindromicPaths (TreeNode root) {
        map=new HashMap<>();
        count=0;
        dfs(root);
        return count;
    }
    public void dfs(TreeNode root)
    {
        if(root==null)return;
        map.put(root.val,map.getOrDefault(root.val,0)+1);
        if(root.left==null&&root.right==null)
        {
        int c=0;
        for(int i:map.keySet())
            if(map.get(i)%2 !=0 ) c++;
        if(c<=1)count++;
        map.put(root.val,map.get(root.val)-1);
        return ;
        }
        dfs(root.left);
        dfs(root.right);
        map.put(root.val,map.get(root.val)-1);
    } 
}

File: January/Python/Pseudo-Palindromic Paths in a Binary Tree.py
class Solution:
    def pseudoPalindromicPaths (self, root: Optional[TreeNode]) -> int:
        count = 0
        # root =[2,3,1,3,1,null,1]
        
        stack =[(root, 0)]
        while stack:
            node, path = stack.pop()
            if node is not None:
                path = path ^ (1 << node.val)
                # if it's a leaf, check if the path ispsuedo-palindromic
                if node.left is None and node.right is None:
                    # check if at most one digits has an odd frequency
                    if path & (path - 1) == 0:
                        count += 1
                else:
                    stack.append((node.left, path))
                    stack.append((node.right, path))

        return count
        
